lib/gc/Dialect/Linalgx/Utils.cpp

```diff
@@ -0,0 +1,515 @@
+  BatchDimMap(ArrayRef<int64_t> batchARef, ArrayRef<int64_t> batchBRef,
+              ArrayRef<int64_t> batchCRef)
+      : batchA(batchARef), batchB(batchBRef), batchC(batchCRef) {}
+  /// Get original arrays
+  ArrayRef<int64_t> getBatchA() const { return ArrayRef<int64_t>(batchA); }
+  ArrayRef<int64_t> getBatchB() const { return ArrayRef<int64_t>(batchB); }
+  ArrayRef<int64_t> getBatchC() const { return ArrayRef<int64_t>(batchC); }
+  /// Get size of batch dims
+  int64_t getBatchNum() const { return batchA.size(); }
+  /// Get attr for batch info
+  bool isBatchEmpty() const {
+    return batchA.empty() && batchB.empty() && batchC.empty();
+  }
+  bool isBatchMatmul() const {
+    return !isBatchEmpty() && batchEqualAB() && batchEqualAC();
+  }
+  bool isBatchReduce() const {
+    return !isBatchEmpty() && batchEqualAB() && batchC.empty();
+  }
+
+private:
+  bool batchEqualAB() const { return llvm::equal(batchA, batchB); }
+  bool batchEqualAC() const { return llvm::equal(batchA, batchC); }
+  SmallVector<int64_t> batchA;
+  SmallVector<int64_t> batchB;
+  SmallVector<int64_t> batchC;
+};
+
+/// PackingMap represent the dim mapping between 2 sets of sorted indices.
+/// PackingMap requires 2 int64 arrays params, it is needed to verify that one
+/// of them contain only 1 index, since multi-dims to multi-dims mapping is not
+/// allowed. This will define a 1->N index set mapping, src is the 1 index, dst
+/// is the multi-dims index list. Some helpers are provided to get the mapping
+/// order(first<-second or first->second) and mapping src/dst indices.
+///
+/// e.g. in A[a,b] -> B[x,y,z], if dim [a] corresponding to dim [x]; dim [b]
+/// corresponding to packed dims [y,z]. We can express it as
+/// `PackingMap<[a] -> [x]>`, `PackingMap<[b] -> [y,z]>`, where
+/// dims mapping order is A -> B
+struct PackingMap {
+public:
+  PackingMap(ArrayRef<int64_t> firstRef, ArrayRef<int64_t> secondRef)
+      : first(firstRef), second(secondRef) {}
+  // Get original arrays
+  ArrayRef<int64_t> getFirst() const { return ArrayRef<int64_t>(first); }
+  ArrayRef<int64_t> getSecond() const { return ArrayRef<int64_t>(second); }
+  // SrcDims.size() == 1; DstDims.size() >= 1
+  ArrayRef<int64_t> getPackingSrcDims() const {
+    return getPackingSrcIndex() == 0 ? getFirst() : getSecond();
+  }
+  ArrayRef<int64_t> getPackingDstDims() const {
+    return getPackingDstIndex() == 0 ? getFirst() : getSecond();
+  }
+  // Index first is 0; Index second is 1
+  unsigned getPackingSrcIndex() const { return getFirst().size() == 1 ? 0 : 1; }
+  unsigned getPackingDstIndex() const { return getFirst().size() == 1 ? 1 : 0; }
+
+private:
+  SmallVector<int64_t> first;
+  SmallVector<int64_t> second;
+};
+
+/// PackingAttr to represent a matmul packing:
+/// vnni or non-vnni matmul, dim size of weight, batch dims, M,N,K packing map
+struct PackingAttr {
+  bool isVnni = false;
+  int64_t weightDims = 0;
+  BatchDimMap batchDimMap;
+  SmallVector<PackingMap> mPacking;
```

Comment:
Can you also state the detailed relation described by `mPacking`, `nPacking`, and `kPacking`? e.g. `mPacking` is input --> output

Reply:
Added